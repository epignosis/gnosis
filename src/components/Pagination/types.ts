import { ExtendableProps } from "types/utils";

export type PaginationProps = ExtendableProps<
  React.HTMLAttributes<HTMLDivElement>,
  {
    current: number;
    totalPages: number;
    list: RowItem[];
    selectionText: string;
    size: number;
    handlePaginationSizeChanged: (size: number) => void;
    handlePaginationNumberChanged: (page: number) => void;
  }
>;

export type RowItem = {
  id: number;
  value: string;
};
