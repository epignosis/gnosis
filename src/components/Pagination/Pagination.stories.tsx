import React from "react";
import { Story } from "@storybook/react";
import Pagination, { PaginationProps } from "./Pagination";

export default {
  component: Pagination,
  title: "Components/Pagination",
  argTypes: {
    onChange: { action: "Changed page!" },
  },
  args: {
    current: 1,
    totalPages: 4,
  },
};

const Template: Story<PaginationProps> = (args) => <Pagination {...args} />;

export const WithNextBtn = Template.bind({});

WithNextBtn.args = {
  current: 1,
};

export const WithBothBtns = Template.bind({});

WithBothBtns.args = {
  current: 2,
  totalPages: 4,
};

export const WithPreviousBtn = Template.bind({});

WithPreviousBtn.args = {
  current: 4,
};
