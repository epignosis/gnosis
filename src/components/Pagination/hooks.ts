import { RefObject, useEffect } from "react";

type Handler = (event: MouseEvent) => void;

// Improved version of https://usehooks.com/useOnClickOutside/
const useClickOutside = (ref: RefObject<HTMLElement>, handler: Handler): void => {
  useEffect(() => {
    let startedInside = false;
    let startedWhenMounted = false;

    const listener = (event: MouseEvent): void => {
      // Do nothing if `mousedown` or `touchstart` started inside ref element
      if (startedInside || !startedWhenMounted) return;
      // Do nothing if clicking ref's element or descendent elements
      if (!ref.current || ref.current.contains(event.target as Node)) return;

      handler(event);
    };

    const validateEventStart = (event: TouchEvent | MouseEvent): void => {
      startedWhenMounted = Boolean(ref.current);
      startedInside = Boolean(ref.current && ref.current.contains(event?.target as Node));
    };

    document.addEventListener("mousedown", validateEventStart);
    document.addEventListener("touchstart", validateEventStart);
    document.addEventListener("click", listener);

    return (): void => {
      document.removeEventListener("mousedown", validateEventStart);
      document.removeEventListener("touchstart", validateEventStart);
      document.removeEventListener("click", listener);
    };
  }, [ref, handler]);
};

export default useClickOutside;
