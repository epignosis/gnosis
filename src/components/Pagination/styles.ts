import { css, SerializedStyles, Theme } from "@emotion/react";
import { hexToRGBA } from "@theme/default/colors";

export const container = ({ pagination }: Theme): SerializedStyles => {
  return css`
    display: flex;
    justify-content: center;
    align-items: center;

    .pagination-options {
      margin: 0 1rem;

      button {
        margin-right: 0.5rem;

        &:last-of-type {
          margin: 0;
        }
      }

      .isActive {
        background: ${hexToRGBA(pagination.background, 0.16)};
        color: ${pagination.color};
      }
    }
  `;
};
