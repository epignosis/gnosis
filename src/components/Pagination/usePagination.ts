import { useMemo } from "react";

export const ellipsis = "...";

const range = (start: number, end: number) => {
  const length = end - start + 1;
  /*
  	Create an array of certain length and set the elements within it from
    start value to end value.
  */
  return Array.from({ length }, (_, idx) => idx + start);
};

const FIRST_PAGE_INDEX = 1; // first page index
const NO_DISPLAYING_BOXES_WITHOUT_ELLIPSIS = 5; // NO_DISPLAYING_BOXES is determined as siblingCount + firstPage + lastPage + currentPage + 2*DOTS

export const usePagination = (
  currentPage: number,
  totalPages: number,
  siblingCount = 1,
): (string | number)[] => {
  return useMemo(() => {
    const totalDisplyedBoxes = siblingCount + NO_DISPLAYING_BOXES_WITHOUT_ELLIPSIS;

    /*
      Case 1:
      If the number of pages is less than the page numbers we want to show in our
      paginationComponent, we return the range [1..totalPages]
    */
    if (totalDisplyedBoxes >= totalPages) {
      return range(FIRST_PAGE_INDEX, totalPages);
    }

    // Calculate left and right sibling index and make sure they are within range 1 and totalPages
    const leftSiblingIndex = Math.max(currentPage - siblingCount, FIRST_PAGE_INDEX);
    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPages);

    // We do not show dots just when there is just one page number to be inserted between
    // the extremes of sibling and the page limits i.e 1 and totalPages.
    // Hence we are using leftSiblingIndex > 2 and rightSiblingIndex < totalPages - 2
    const shouldShowLeftDots = leftSiblingIndex > 2;
    const shouldShowRightDots = rightSiblingIndex < totalPages - 2;

    /*
    	Case 2: No left dots to show, but rights dots to be shown
    */
    if (!shouldShowLeftDots && shouldShowRightDots) {
      const leftItemCount = 3 + 2 * siblingCount;
      const leftRange = range(1, leftItemCount);

      return [...leftRange, ellipsis, totalPages];
    }

    /*
    	Case 3: No right dots to show, but left dots to be shown
    */
    if (shouldShowLeftDots && !shouldShowRightDots) {
      const rightItemCount = 3 + 2 * siblingCount;
      const rightRange = range(totalPages - rightItemCount + 1, totalPages);
      return [FIRST_PAGE_INDEX, ellipsis, ...rightRange];
    }

    /*
    	Case 4: Both left and right dots to be shown
    */
    if (shouldShowLeftDots && shouldShowRightDots) {
      const middleRange = range(leftSiblingIndex, rightSiblingIndex);
      return [FIRST_PAGE_INDEX, ellipsis, ...middleRange, ellipsis, totalPages];
    }

    return [];
  }, [currentPage, totalPages]);
};
