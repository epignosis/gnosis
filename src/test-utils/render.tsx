import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
