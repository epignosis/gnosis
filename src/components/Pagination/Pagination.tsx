import React, { FC } from "react";
import classNames from "classnames";
import Button from "../Button/Button";
import { ArrowLeftSVG, ArrowRightSVG } from "../../icons/";
import { container } from "./styles";
import { usePagination, ellipsis } from "./usePagination";

export type PaginationProps = {
  current: number;
  totalPages: number;
  onChange: (page: number) => void;
  containerAttrs?: React.HTMLAttributes<HTMLDivElement>;
};

const Pagination: FC<PaginationProps> = ({ current, onChange, totalPages, containerAttrs }) => {
  const paginationRange = usePagination(current, totalPages);

  const classNamesContainer = (pageNumber: number) =>
    classNames({
      isActive: pageNumber === current,
    });

  return (
    <div css={container} {...containerAttrs}>
      <Button
        className="previous-page-btn"
        data-testid="previous-page-btn"
        name="Previous page"
        onClick={(): void => onChange(current - 1)}
        color="secondary"
        noGutters
        disabled={current === 1}
      >
        <ArrowLeftSVG height={22} />
      </Button>

      <div className="pagination-options">
        {paginationRange.map((pageNumber) => {
          if (pageNumber === ellipsis) {
            return ellipsis;
          }

          return (
            <Button
              type="button"
              key={pageNumber}
              id={pageNumber.toString()}
              data-testid="pagination-page"
              onClick={(): void => onChange(pageNumber as number)}
              variant="ghost"
              noGutters
              className={classNamesContainer(pageNumber as number)}
            >
              {pageNumber}
            </Button>
          );
        })}
      </div>

      <Button
        className="next-page-btn"
        data-testid="next-page-btn"
        name="Next page"
        onClick={(): void => onChange(current + 1)}
        color="secondary"
        noGutters
        disabled={current === totalPages}
      >
        <ArrowRightSVG height={22} />
      </Button>
    </div>
  );
};

export default Pagination;
