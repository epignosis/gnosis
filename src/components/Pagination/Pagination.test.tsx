import React from "react";
import Pagination from "./Pagination";
import { render, screen } from "@test-utils/render";

describe("<Pagination />", () => {
  it("renders", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={1} totalPages={1} onChange={mockFn} />);

    expect(container).toBeInTheDocument();
    expect(container).toBeVisible();
  });

  it("has 2 pages", () => {
    const mockFn = jest.fn();
    render(<Pagination current={1} totalPages={2} onChange={mockFn} />);

    const pageOptions = screen.getAllByRole("option");

    expect(pageOptions).toHaveLength(2);
  });

  it("doesn't have a next page", () => {
    const mockFn = jest.fn();
    const { queryByText } = render(<Pagination current={2} totalPages={2} onChange={mockFn} />);

    const nextBtn = queryByText(/Next/);

    expect(nextBtn).toBeNull();
  });

  it("doesn't have a previous page", () => {
    const mockFn = jest.fn();
    const { queryByText } = render(<Pagination current={1} totalPages={2} onChange={mockFn} />);

    const previousBtn = queryByText(/Previous/);

    expect(previousBtn).toBeNull();
  });

  it("has next & previous pages", () => {
    const mockFn = jest.fn();
    const { queryByText } = render(<Pagination current={2} totalPages={4} onChange={mockFn} />);

    const nextBtn = queryByText(/Next/);
    const previousBtn = queryByText(/Previous/);

    expect(nextBtn).not.toBeNull();
    expect(previousBtn).not.toBeNull();
  });

  it("matches snapshot with on page", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={1} totalPages={1} onChange={mockFn} />);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with 2 pages", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={1} totalPages={2} onChange={mockFn} />);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with no previous page", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={1} totalPages={2} onChange={mockFn} />);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with no next page", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={2} totalPages={2} onChange={mockFn} />);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with previous and next pages", () => {
    const mockFn = jest.fn();
    const { container } = render(<Pagination current={2} totalPages={4} onChange={mockFn} />);

    expect(container).toMatchSnapshot();
  });
});
