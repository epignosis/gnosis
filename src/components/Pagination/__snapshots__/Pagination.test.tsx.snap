// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> matches snapshot with on page 1`] = `
<div>
  <div
    class="html-class css-1de41tm-container-container"
    id="my-id"
  >
    <button
      class="previous-page-btn solid css-hhqmb3-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
      data-testid="previous-page-btn"
      name="Previous page"
    >
      <span
        class="btn-text"
      >
        <svg-file-stub
          height="22"
        />
      </span>
    </button>
    <div
      class="pagination-options"
    >
      <button
        class="ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="1"
        type="button"
      >
        <span
          class="btn-text"
        >
          1
        </span>
      </button>
      <button
        class="isActive ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="2"
        type="button"
      >
        <span
          class="btn-text"
        >
          2
        </span>
      </button>
      <button
        class="ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="3"
        type="button"
      >
        <span
          class="btn-text"
        >
          3
        </span>
      </button>
      <button
        class="ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="4"
        type="button"
      >
        <span
          class="btn-text"
        >
          4
        </span>
      </button>
      <button
        class="ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="5"
        type="button"
      >
        <span
          class="btn-text"
        >
          5
        </span>
      </button>
      <button
        class="ghost css-1030g3v-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
        data-testid="pagination-page"
        id="6"
        type="button"
      >
        <span
          class="btn-text"
        >
          6
        </span>
      </button>
    </div>
    <button
      class="next-page-btn solid css-hhqmb3-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
      data-testid="next-page-btn"
      name="Next page"
    >
      <span
        class="btn-text"
      >
        <svg-file-stub
          height="22"
        />
      </span>
    </button>
  </div>
</div>
`;
